### PR TITLE
dts: gpio: Add Microchip XEC GPIO macros for use in device tree

### PR DESCRIPTION
**Pull request by jvasanth1**
For Microchip XEC devices, create a set of defines for all GPIO pins that specify the DT GPIO bank name used in the chip level DTSI files and the bit position in that bank.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

### CODE REVIEW

**Review by vphirema — DISMISSED** [review]

**Comment by jvasanth1** [comment]
@albertofloyd can you help to review?

**Review by vphirema — DISMISSED** [review]

**Review by cfriedt — COMMENTED** [review]
@jvasanth1 - are there be other related Microchip SoCs that define the same GPIOs with different banks / parameters? I'm just trying to understand the main reason for adding them.

**Comment by jvasanth1** [comment]
> @jvasanth1 - are there be other related Microchip SoCs that define the same GPIOs with different banks / parameters? I'm just trying to understand the main reason for adding them.

MEC15xx and MEC172x family of SoCs uses this octal numbering to organize pins in banks of 32. The main reason for adding these changes was to easily specify the GPIO using their octal number instead of a bank name and the position in the bank.

**Review by cfriedt — DISMISSED** [review]

**Review by gmarull — CHANGES_REQUESTED** [review]
We don't have any other platform _hiding_ port/pin using macros. Something like this was dismissed for pinctrl, so I think we need at least to discuss this. cc: @galak @mbolivar-nordic .

**Comment by jvasanth1** [comment]
> We don't have any other platform _hiding_ port/pin using macros. Something like this was dismissed for pinctrl, so I think we need at least to discuss this. cc: @galak @mbolivar-nordic .

cc @vphirema as well

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
I agree with @gmarull. This needs further discussion. @carlescufi Perhaps a candidate for the Architecture WG meeting?

**Comment by vphirema** [comment]
cc: @keith-zephyr

**Review by mbolivar-nordic — COMMENTED** [review]
I don't like this either, FTR

**Comment by keith-zephyr** [comment]
On a HW schematic. the GPIO pins are specified in groups of 8, using octal counting.  `GPIO000` -`GPIO007`, `GPIO010` - `GPIO017`, `GPIO020` - `GPIO027`, etc. However, the GPIO input/output registers manage GPIOs in groups of 32.

I see this change as exposing the HW level port/pin information to the user.  Without this change, to use `GPIO070`, the user had to specify `gpios = <&gpio_040_076 24 GPIO_INPUT>;`.  The "port", `gpio_040_076`, is easy enough to determine, but the "pin" is not obvious.  The user has to convert the GPIO number from octal to decimal (70 -> 56) and then mod by 32 to get the correct bit position.

**Comment by mbolivar-nordic** [comment]
The main thing that bothers me about this is the use of a macro that expands to 2 tokens within a 3-token phandle+specifier block within the DTS. If this is really a usability win, why not put the flags cell into the macro arguments too, so you'll get DTS which looks like `<MY_SOC_GPIO_SPEC(soc_specific_num1, flags1)>, <MY_SOC_GPIO_SPEC(soc-specific_num2, flags2)>, ...`?

That feels a lot more readable to me, at least. It's more obvious how this relates to the way GPIOs are specified in DTS. If the whole point is readability / usability, the current half-way approach doesn't seem to go far enough.

**Comment by jvasanth1** [comment]
> The main thing that bothers me about this is the use of a macro that expands to 2 tokens within a 3-token phandle+specifier block within the DTS. If this is really a usability win, why not put the flags cell into the macro arguments too, so you'll get DTS which looks like `<MY_SOC_GPIO_SPEC(soc_specific_num1, flags1)>, <MY_SOC_GPIO_SPEC(soc-specific_num2, flags2)>, ...`?
> 
> That feels a lot more readable to me, at least. It's more obvious how this relates to the way GPIOs are specified in DTS. If the whole point is readability / usability, the current half-way approach doesn't seem to go far enough.

can you be more specific how the macro function would look like? so if I understand correctly, we want some macro function XEC_GPIO_SPEC to be called as follows

`XEC_GPIO_SPEC(156, GPIO_ACTIVE_LOW)`

and this should result in

`&gpio_140_176 14 GPIO_ACTIVE_LOW`

**Comment by mbolivar-nordic** [comment]
@jvasanth1 something like that, yes.

**Comment by henrikbrixandersen** [comment]
The Linux kernel has a few examples of similar macros, but they do not include the GPIO flags as macro arguments. Examples:
- `RZG2L_GPIO()`: https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/pinctrl/renesas,rzg2l-pinctrl.yaml
- `TEGRA_GPIO()`: https://elixir.bootlin.com/linux/latest/source/include/dt-bindings/gpio/tegra-gpio.h

**Comment by jvasanth1** [comment]
we are proposing the following approach
_microchip-xec-gpio.h_ will have following definitions

```
#define DT_PASTER(x, y, z)    x y z

#define MCHP_GPIO156(x)    DT_PASTER(&gpio_140_176, 14, x)
```
and so on for each GPIO

the board level dts and overlays can use these macros as follows:

`gpios = MCHP_GPIO156(GPIO_ACTIVE_LOW)`

Please let us know if this approach is acceptable. We tried to derive the GPIO bank and bit position in the bank using octal arithmetic but looks like it is not possible to do so with macro preprocessing

**Comment by mbolivar-nordic** [comment]
> The Linux kernel has a few examples of similar macros, but they do not include the GPIO flags as macro arguments.

I think this is a convincing argument, please disregard my previous comment.

**Comment by mbolivar-nordic** [comment]
> #define DT_PASTER(x, y, z)    x y z

Please note that the `DT_` namespace is reserved for `zephyr/devicetree.h` and its sub-headers. Defining a macro that begins with `DT_` outside of these headers or their transitive includes is not acceptable.

**Comment by jvasanth1** [comment]
> > `#define DT_PASTER(x, y, z)    x y z`
> 
> Please note that the `DT_` namespace is reserved for `zephyr/devicetree.h` and its sub-headers. Defining a macro that begins with `DT_` outside of these headers or their transitive includes is not acceptable.

Okay, will define it as XEC_PASTER

and proceed to update the PR as described in https://github.com/zephyrproject-rtos/zephyr/pull/53109#issuecomment-1398720814

**Comment by mbolivar-nordic** [comment]
> and proceed to update the PR as described in [#53109 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/53109#issuecomment-1398720814)

It's fine for me if you want to go that way, but please note https://github.com/zephyrproject-rtos/zephyr/pull/53109#issuecomment-1399041930

**Comment by jvasanth1** [comment]
please help review and approve

**Review by vphirema — DISMISSED** [review]

File: include/zephyr/dt-bindings/gpio/microchip-xec-gpio.h [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+ * Microchip XEC documentation uses octal GPIO pin
+ * numbering. These macros do not require the user to do octal arithmetic
+ * to derive the pin's bit postion.
+ *
+ * Example DT usage: gpios = <&MCHP_GPIO176 GPIO_ACTIVE_HIGH>;
+ *
+ * @{
+ */
+
+#define XEC_PASTER(x, y, z)	x y z
```
Comment (mbolivar-nordic) [inline include/zephyr/dt-bindings/gpio/microchip-xec-gpio.h:21]: minor point: on a second look, I think this should be named XEC_GPIO_HELPER. "Pasting" in the C preprocessor almost always refers to usage of the `##` operator, which this is not using, so the name is misleading in my opinion. This seems reasonable to me though.

File: include/zephyr/dt-bindings/gpio/microchip-xec-gpio.h [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+#define ZEPHYR_INCLUDE_DT_BINDINGS_GPIO_MICROCHIP_XEC_GPIO_H_
+
+/**
+ * @brief Microchip XEC GPIO bank and bit position convenience defines
+ *
+ * Microchip XEC documentation uses octal GPIO pin
+ * numbering. These macros do not require the user to do octal arithmetic
+ * to derive the pin's bit postion.
+ *
+ * Example DT usage: gpios = <&MCHP_GPIO176 GPIO_ACTIVE_HIGH>;
```
[the remark this answers is not in the document]
Comment (jvasanth1) [reply include/zephyr/dt-bindings/gpio/microchip-xec-gpio.h]: updated

File: include/zephyr/dt-bindings/gpio/microchip-xec-gpio.h (hunk 1 above)
Comment (jvasanth1) [reply include/zephyr/dt-bindings/gpio/microchip-xec-gpio.h]: updated to use XEC_GPIO_HELPER

**Review by mbolivar-nordic — DISMISSED** [review]

**Review by vphirema — DISMISSED** [review]

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Was it intentional to keep passing the "flags" to the helper macro?

**Comment by jvasanth1** [comment]
> Was it intentional to keep passing the "flags" to the helper macro?

Yes, so that the macro expands to 3 tokens as required within the DTS

**Comment by jvasanth1** [comment]
> @jvasanth1 I'm not sure where you are getting the notion that this is "required". As I wrote above and reminded you about already in my previous review, the current approach is OK with me, but it is not a requirement given that @henrikbrixandersen found counterexamples in Linux. Please re-read the messages in scrollback.

Ok

**Review by cfriedt — DISMISSED** [review]

**Comment by henrikbrixandersen** [comment]
> The Linux kernel has a few examples of similar macros, but they do not include the GPIO flags as macro arguments. Examples:
> 
> * `RZG2L_GPIO()`: https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/pinctrl/renesas,rzg2l-pinctrl.yaml
> * `TEGRA_GPIO()`: https://elixir.bootlin.com/linux/latest/source/include/dt-bindings/gpio/tegra-gpio.h

Please follow the macro convention used above where the `flags` is not passed to the helper macro.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Please see the requested change above.

**Comment by jvasanth1** [comment]
> Please see the requested change above.

we reviewed the examples but our use case is not straightforward

we are expecting a macro of type `XEC_GPIO_SPEC(156)` to turn into `&gpio_140_176 14`

We tried to derive the GPIO bank and bit position in the bank using octal arithmetic but looks like it is not possible to do so with macro preprocessing

**Comment by henrikbrixandersen** [comment]
> > Please see the requested change above.
> 
> we reviewed the examples but our use case is not straightforward
> 
> we are expecting a macro of type `XEC_GPIO_SPEC(156)` to turn into `&gpio_140_176 14`
> 
> We tried to derive the GPIO bank and bit position in the bank using octal arithmetic but looks like it is not possible to do so with macro preprocessing

I don't see any issues in that. You are just passing the GPIO flags straight through the macro (`GPIO_ACTIVE_HIGH` and friends are passed as `x`). 

Please also rename the macro arguments to something more descriptive that x, y, and z.

File: include/zephyr/dt-bindings/gpio/microchip-xec-gpio.h [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+ *
+ * Example DT usage: gpios = <MCHP_GPIO176(GPIO_ACTIVE_HIGH)>;
+ *
+ * @{
+ */
+
+#define XEC_GPIO_HELPER(x, y, z)	x y z
+
+/* bank A */
+#define MCHP_GPIO000(x) XEC_GPIO_HELPER(&gpio_000_036, 0, x)
```
Comment (henrikbrixandersen) [inline include/zephyr/dt-bindings/gpio/microchip-xec-gpio.h:24]: Something along the lines of:

```suggestion
/* bank A */
#define MCHP_GPIO000 XEC_GPIO_HELPER(&gpio_000_036, 0)
```

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

**Comment by albertofloyd** [comment]
@jvasanth1 overall I don't like this either, but it seems after feedback from device tree experts this going in a good direction. 

Reading the comments I'm just wondering if a simpler approach to tackle the gpio assignment complexity in device tree was ever considered. 

`e.g. current 

- gpios = <&gpio_140_176 **15** GPIO_ACTIVE_HIGH>;  
+ gpios = <&gpio_140_176 **MCHP_GPIO157** GPIO_ACTIVE_HIGH>;  
`
I agree it's hard to tell which pin is referred by 15, but wouldn't it be simpler to use to existing macros soc/arm/microchip_mec/common/soc_pins.h  instead of defining new ones? 

In the same line of thought, how to avoid the potential clash between existing macros in said file and new ones with exactly.
Is it guaranteed

**Comment by rgundi** [comment]
I think it would look better if you adopt @henrikbrixandersen's [suggestion](https://github.com/zephyrproject-rtos/zephyr/pull/53109#discussion_r1110184843). That said, this macro can be a bit confusing as we already have something like "#define MCHP_GPIO_000   (0U)" in soc/arm/microchip_mec/common/soc_pins.h.

**Comment by jvasanth1** [comment]
@albertofloyd in the simple approach we still need to specify the bank (or port) so I am okay to proceed with @henrikbrixandersen [suggestion](https://github.com/zephyrproject-rtos/zephyr/pull/53109#discussion_r1110184843) - also echoed by @rgundi

**Comment by albertofloyd** [comment]
> ntial macro name clash, I will name the macros as follo

Agreed @jvasanth1

**Comment by jvasanth1** [comment]
any suggestion to fix the failing compliance check

```
Warning, treated as error:
/home/runner/work/zephyr/zephyr/include/zephyr/dt-bindings/gpio/microchip-xec-gpio.h:16: Unsupported xml/html tag <MCHP_GPIO_DECODE_176> found
FAILED: CMakeFiles/html 
```

File: include/zephyr/dt-bindings/gpio/microchip-xec-gpio.h [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+#define ZEPHYR_INCLUDE_DT_BINDINGS_GPIO_MICROCHIP_XEC_GPIO_H_
+
+/**
+ * @brief Microchip XEC GPIO bank and bit position convenience defines
+ *
+ * Microchip XEC documentation uses octal GPIO pin
+ * numbering. These macros do not require the user to do octal arithmetic
+ * to derive the pin's bit postion.
+ *
+ * Example DT usage: gpios = <MCHP_GPIO_DECODE_176 GPIO_ACTIVE_HIGH>;
```
Comment (gmarull) [inline include/zephyr/dt-bindings/gpio/microchip-xec-gpio.h:16]: if embedding sample code please do this:

```suggestion
 * Example DT usage:
 *
 * @code{.dts}
 * gpios = <MCHP_GPIO_DECODE_176 GPIO_ACTIVE_HIGH>;
 * @endcode
```

Comment (jvasanth1) [reply include/zephyr/dt-bindings/gpio/microchip-xec-gpio.h]: Thanks @gmarull

**Review by gmarull — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Review by rgundi — APPROVED** [review]

**Pull request merged**